libs/vkd3d/cache.c

```diff
@@ -118,12 +118,125 @@ VkResult vkd3d_serialize_pipeline_state(const struct d3d12_pipeline_state *state
     return VK_SUCCESS;
 }
 
+struct vkd3d_cached_pipeline_key
+{
+    size_t name_length;
+    const void *name;
+};
+
+struct vkd3d_cached_pipeline_data
+{
+    size_t blob_length;
+    const void *blob;
+    bool is_new;
+};
+
+struct vkd3d_cached_pipeline_entry
+{
+    struct hash_map_entry entry;
+    struct vkd3d_cached_pipeline_key key;
+    struct vkd3d_cached_pipeline_data data;
+};
+
+uint32_t vkd3d_cached_pipeline_hash(const void *key)
+{
+    const struct vkd3d_cached_pipeline_key *k = key;
+    uint32_t hash = 0;
+    size_t i;
+
+    for (i = 0; i < k->name_length; i += 4)
+    {
+        uint32_t accum = 0;
+        memcpy(&accum, (const char*)k->name + i,
+                min(k->name_length - i, sizeof(accum)));
+        hash = hash_combine(hash, accum);
+    }
+
+    return hash;
+}
+
+bool vkd3d_cached_pipeline_compare(const void *key, const struct hash_map_entry *entry)
+{
+    const struct vkd3d_cached_pipeline_entry *e = (const struct vkd3d_cached_pipeline_entry*)entry;
+    const struct vkd3d_cached_pipeline_key *k = key;
+
+    return k->name_length == e->key.name_length &&
+            !memcmp(k->name, e->key.name, k->name_length);
+}
+
+struct vkd3d_serialized_pipeline
+{
+    uint32_t name_length;
+    uint32_t blob_length;
+    char data[];
+};
+
+#define VKD3D_PIPELINE_LIBRARY_VERSION 1
+
+struct vkd3d_serialized_pipeline_library
+{
+    uint32_t version;
```

Comment:
A more specialized magic number would be nice.

Reply:
See above, fixed.